boards/esp32-ethernet-kit/include/periph_conf.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2018 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     boards_esp32_esp-ethernet-kit
+ * @brief       Peripheral MCU configuration for Espressif ESP32-Ethernet-Kit
+ * @file
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ */
+
+#ifndef PERIPH_CONF_H
+#define PERIPH_CONF_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+ extern "C" {
+#endif
+
+/**
+ * @name    ADC and DAC channel configuration
+ * @{
+ */
+
+/**
+ * @brief   Declaration of GPIOs that can be used as ADC channels
+ *
+ * @note As long as the GPIOs listed in ADC_GPIOS are not initialized as ADC
+ * channels with the ```adc_init``` function, they can be used for other
+ * purposes.
+ */
+#ifndef ADC_GPIOS
+#define ADC_GPIOS   { GPIO32, GPIO33, GPIO34, GPIO35, GPIO36, GPIO39 }
+#endif
+
+/**
+ * @brief   Declaration of GPIOs that can be used as DAC channels
+ *
+ * ESP-WROVER-KIT has no GPIOs left that might be used as DAC channels.
+ */
+#ifndef DAC_GPIOS
+#define DAC_GPIOS   { }
+#endif
+/** @} */
+
+
+/**
+ * @name   PWM channel configuration
+ *
+ * @note As long as the according PWM device is not initialized with
+ * the ```pwm_init```, the GPIOs declared for this device can be used
+ * for other purposes.
+ *
+ * @{
+ */
+#ifndef PWM0_GPIOS
+#define PWM0_GPIOS  { }
```

Comment:
As I understand the [GPIO allocation table](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/esp32/get-started-ethernet-kit.html#gpio-allocation-summary), GPIO2 and GPIO4 are not used. You could define them as PWM channels.
```suggestion
#define PWM0_GPIOS  { GPIO2, GPIO4 }
```

Reply:
Added these and all pins from ADC above. Saw that ADC is only available on 2 and 4 of them and updated that.

Reply:
Ups, no. GPIOs 32 ... 39 are only inputs. You can't define them as PWM pins. And of course, the are ADC pins. Please change it back.

Reply:
GPIO32 and 33 seem to be I/O according to https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf ?

Is it ok to share pins between PWM and ADC or should I put two in each?

Reply:
> GPIO32 and 33 seem to be I/O according to https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf ?

Yes, of course. GPIOs 34 ... 39 are a bit special. They can't be used as outputs and they don't have software controllable pullups/pulldowns. So their function is limited to mode `GPIO_IN` and ADC channels.

There is another restriction. The ESP32 has 2 ADCs. Since ADC2 is not available when using the WiFi interface, the GPIOs connected to ADC2 cannot be used as ADC channels, see [ADC Channels](http://doc.riot-os.org/group__cpu__esp32.html#esp32_adc_channels) section in ESP32 documentation of RIOT:
- ADC1: GPIOs 32-39
- ADC2: GPIOs 0, 2, 4, 12-15, 25-27

Having all these restrictions in mind, I would only use GPIOs in ADC/I2C/PWM/SPI configurations that are broken out (these are GPIOs 2, 4, 32, 33, 34, 35, 36, 39) and are not affected by these limitations. A board definition should define an I2C interface. So I would suggest to define

- GPIO32 as `I2C0_SDA` pin and GPIO33 as `I2C0_SCL pin`.
- GPIO34, GIO33, GPIO35, GPIO36 and GPIO39 as `ADC_GPIOS`
- GPIO2 and GPIO4 as `PWM0_GPIOS`

```c
#ifndef ADC_GPIOS
#define ADC_GPIOS   { GPIO33, GPIO34, GPIO35, GPIO36, GPIO39 }
#endif

#ifndef I2C0_SPEED
#define I2C0_SPEED  I2C_SPEED_FAST  /**< I2C bus speed of I2C_DEV(0) */
#endif
#ifndef I2C0_SCL
#define I2C0_SCL    GPIO33          /**< SCL signal of I2C_DEV(0) */
#endif
#ifndef I2C0_SDA
#define I2C0_SDA    GPIO32          /**< SDA signal of I2C_DEV(0) */
#endif

#ifndef PWM0_GPIOS
#define PWM0_GPIOS  { GPIO2, GPIO4 }
#endif
```

You might also think about to use the JTAG pins GPIO12, GPIO13, GPIO14, GPIO15 in definitions with a clear statement in the documentation that these GPIOs are only available, if the function switches for the JTAG signals are OFF. To have an information at compile time whether the JTAG interface is used, you could define a pseudomodule, lets say `esp_jtag`. If this module is used, the function switches for JTAG signals are ON and GPIO12, GPIO13, GPIO14, GPIO15 are used for JTAG debugging. If the module is not used, you could define SPI0 instead for these GPIOS, e.g.
```c
#ifndef MODULE_ESP_JTAG
#define SPI0_CTRL   HSPI    /**< HSPI is used as SPI_DEV(0) */
#define SPI0_SCK    GPIO14  /**< HSPI SCK */
#define SPI0_MISO   GPIO12  /**< HSPI MISO */
#define SPI0_MOSI   GPIO13  /**< HSPI MOSI */
#define SPI0_CS0    GPIO15  /**< HSPI CS0 */
#endif
```
Don't forget to copy the documentation for these configurations
 
> Is it ok to share pins between PWM and ADC or should I put two in each?

It would be possible to use the same pins in both definitions as long as these functions are used exclusively. But with the configuration above, it wouldn't be necessary.

Reply:
Updated the ADC/I2C/PWM sections.

I prefer to not do the SPI/JTAG change right now, would be better as a follow-up change I think.

Reply:
> * GPIO34, GIO33, GPIO35, GPIO36 and GPIO39 as `ADC_GPIOS`
> 
> ```diff
> -#define ADC_GPIOS   { GPIO33, GPIO34, GPIO35, GPIO36, GPIO39 }
> +#define ADC_GPIOS   { GPIO34, GPIO35, GPIO36, GPIO39 }

Sorry, there were a small mistake in ADC channel list. GPIO33 is occupied by I2C.

> I prefer to not do the SPI/JTAG change right now, would be better as a follow-up change I think.

Why? It's just a little change. It's much easier to get it merged it with the initial commit than with a follow-up. SPI is an important peripheral interface that should be offered from the beginning if possible.

Reply:
> I prefer to not do the SPI/JTAG change right now, would be better as a follow-up change I think.

Furthermore, JTAG is only required during development, if it is necessary to debug. If the board is in operation, having an SPI might be more impartant than having a JTAG.

Reply:
Defining a pseudomodule is pretty easy. Just add
```
PSEUDOMODULES += esp_jtag
```
to `cpu/esp32/Makfile.include`.

Having this module, it would also be possible to define further dependencies. For example, defining
```
ifneq (,$(filter esp_jtag,$(USEMODULE)))
  FEATURES_REQUIRED += esp_jtag
endif
```
in `cu/esp32/Makefile.include` would require that the board has a JTAG on board. Each board that provides the JTAG on-board (`esp32-ethernet-kit`, `esp32-wrover-kit`) could then define
```
FEATURES_PROVIDED += esp_jtag
```
Finally, for other boards the pin configuration could also be defined depending on whether `esp_jtag` is enabled and thus the JTAG pins cannot be used for other purposes.

I haven't introduced this pseudo module yet, since JTAG is not optional for the `esp32-wrover-kit' board. But I think now is the time to do it.

Reply:
Ok, thanks for the help, I have added the esp_jtag module and feature.

The esp32-wrover-kit has jumpers to disable JTAG at least in v4.1:
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/esp32/get-started-wrover-kit.html#setup-options